packages/mdc-select/index.ts

```diff
@@ -0,0 +1,596 @@
+        this.selectedText_!.setAttribute('aria-disabled', isDisabled.toString());
+        if (this.hiddenInput_) {
+          this.hiddenInput_.disabled = isDisabled;
+        }
+      },
+      checkValidity: () => {
+        const classList = this.root_.classList;
+        if (classList.contains(cssClasses.REQUIRED) && !classList.contains(cssClasses.DISABLED)) {
+          // See notes for required attribute under https://www.w3.org/TR/html52/sec-forms.html#the-select-element
+          // TL;DR: Invalid if no index is selected, or if the first index is selected and has an empty value.
+          return this.selectedIndex !== -1 && (this.selectedIndex !== 0 || Boolean(this.value));
+        } else {
+          return true;
+        }
+      },
+      setValid: (isValid: boolean) => {
+        this.selectedText_!.setAttribute('aria-invalid', (!isValid).toString());
+        if (isValid) {
+          this.root_.classList.remove(cssClasses.INVALID);
+        } else {
+          this.root_.classList.add(cssClasses.INVALID);
+        }
+      },
+    };
+    // tslint:enable:object-literal-sort-keys
+  }
+
+  private getCommonAdapterMethods_() {
+    // tslint:disable:object-literal-sort-keys
+    return {
+      addClass: (className: string) => this.root_.classList.add(className),
+      removeClass: (className: string) => this.root_.classList.remove(className),
+      hasClass: (className: string) => this.root_.classList.contains(className),
+      setRippleCenter: (normalizedX: number) => this.lineRipple_ && this.lineRipple_.setRippleCenter(normalizedX),
+      activateBottomLine: () => this.lineRipple_ && this.lineRipple_.activate(),
+      deactivateBottomLine: () => this.lineRipple_ && this.lineRipple_.deactivate(),
+      notifyChange: (value: string) => {
+        const index = this.selectedIndex;
+        this.emit<SelectEventDetail>(strings.CHANGE_EVENT, {value, index}, true /* shouldBubble  */);
+      },
+    };
+    // tslint:enable:object-literal-sort-keys
+  }
+
+  private getOutlineAdapterMethods_() {
+    // tslint:disable:object-literal-sort-keys
+    return {
+      hasOutline: () => Boolean(this.outline_),
+      notchOutline: (labelWidth: number) => this.outline_ && this.outline_.notch(labelWidth),
+      closeOutline: () => this.outline_ && this.outline_.closeNotch(),
+    };
+    // tslint:enable:object-literal-sort-keys
+  }
+
+  private getLabelAdapterMethods_() {
+    return {
+      floatLabel: (shouldFloat: boolean) => this.label_ && this.label_.float(shouldFloat),
+      getLabelWidth: () => this.label_ ? this.label_.getWidth() : 0,
+    };
+  }
+
+  /**
+   * Calculates where the line ripple should start based on the x coordinate within the component.
+   */
+  private getNormalizedXCoordinate_(evt: MouseEvent): number {
+    const targetClientRect = (evt.target as Element).getBoundingClientRect();
+    const xCoordinate = evt.clientX; // TODO(acdvorak): How should this be typed?
```

Comment:
isn't this inferred to be a number already?

Reply:
The problem is that this method actually takes a `MouseEvent | TouchEvent`, and `clientX` isn't defined on `TouchEvent`. I'm actually not sure how this code works at all right now.

```ts
type PointerEventType = 'mousedown' | 'touchstart';

const POINTER_EVENTS: PointerEventType[] = ['mousedown', 'touchstart'];

class MDCSelect {
  private handleClick_!: SpecificEventListener<PointerEventType>; // assigned in initialize()

  initialSyncWithDOM() {
    this.handleClick_ = (evt) => {
      if (this.selectedText_) {
        this.selectedText_.focus();
      }
      this.foundation_.handleClick(this.getNormalizedXCoordinate_(evt));
    };

    POINTER_EVENTS.forEach((evtType) => {
      element.addEventListener(evtType, this.handleClick_ as EventListener);
    });
  }

  private getNormalizedXCoordinate_(evt: MouseEvent | TouchEvent): number {
    const targetClientRect = (evt.target as Element).getBoundingClientRect();
    const xCoordinate = evt.clientX; // TODO(acdvorak): How should this be typed?
    return xCoordinate - targetClientRect.left;
  }
}
```

Reply:
OH! Wait I think there is a bug for this...I don't think it works.

Reply:
Having a hard time finding it, but I do believe this was an issue in either React or Web or both.

Reply:
https://github.com/material-components/material-components-web-react/issues/275#issuecomment-424272414

Reply:
Didn't we fix this somewhere to dig out `touches[0]`, which does have clientX?

Reply:
The above issue I linked to is text field. So I think we also need to fix for select.

Reply:
Good call! A [type guard](https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards) should do the trick:

```ts
class MDCSelect {
  private getNormalizedXCoordinate_(evt: MouseEvent | TouchEvent): number {
    const targetClientRect = (evt.target as Element).getBoundingClientRect();
    const xCoordinate = this.isTouchEvent_(evt) ? evt.touches[0].clientX : evt.clientX;
    return xCoordinate - targetClientRect.left;
  }

  private isTouchEvent_(evt: MouseEvent | TouchEvent): evt is TouchEvent {
    return Boolean((evt as TouchEvent).touches);
  }
}
```

Reply:
Done